test/gateway.e2e-spec.ts

```diff
@@ -0,0 +1,133 @@
+import { INestApplication, Type, WebSocketAdapter } from '@nestjs/common';
+import { IoAdapter } from '@nestjs/platform-socket.io';
+import { WsAdapter } from '@nestjs/platform-ws';
+import { Test } from '@nestjs/testing';
+import * as Io from 'socket.io-client';
+import * as WebSocket from 'ws';
+import { GatewayModule } from './app/gateways/gateway.module';
+import { createConnection, wsClose, wsPromise } from './utility/ws-promise';
+
+describe.each`
+  adapter      | server         | client                                 | protocol  | sendMethod | serializer                                                 | deserializer
+  ${IoAdapter} | ${'Socket.io'} | ${(url: string) => Io(url)}            | ${'http'} | ${'emit'}  | ${(message: string) => message}                            | ${(message: string) => JSON.parse(message)}
+  ${WsAdapter} | ${'Websocket'} | ${(url: string) => new WebSocket(url)} | ${'ws'}   | ${'send'}  | ${(message: string) => JSON.stringify({ event: message })} | ${(message: string) => message}
+`(
```

Comment:
Damn did not know about this, awesome :D

Reply:
[You should check out my tests here] (https://github.com/jmcdo29/ogma/blob/master/integration/test/rpc.spec.ts) for my Ogma integration tests. Those were fun to write. The power of `describe.each` and `it.each`. Though, ws and SocketIO get a little funky with the bed if the `client` `serializer` and `deserializer`